src/toast/ops/totalconvolve.py

```diff
@@ -0,0 +1,654 @@
+        """
+        if det not in focalplane:
+            raise RuntimeError(f"focalplane does not include {det}")
+        props = focalplane[det]
+        if "psi_pol" in props.colnames:
+            psi_pol = props["pol_angle"].to_value(u.radian)
+        elif "pol_angle" in props.colnames:
+            warnings.warn(
+                "Use psi_pol and psi_uv rather than pol_angle", DeprecationWarning
+            )
+            psi_pol = props["pol_angle"].to_value(u.radian)
+        else:
+            raise RuntimeError(f"focalplane[{det}] does not include psi_pol")
+        return psi_pol
+
+    def _get_psi_uv(self, focalplane, det):
+        """Parse Pxx basis angle in radians from the focalplane
+        dictionary.  The angle is measured from Dxx to Pxx basis.
+        """
+        if det not in focalplane:
+            raise RuntimeError(f"focalplane does not include {det}")
+        props = focalplane[det]
+        if "psi_uv_deg" in props.colnames:
+            psi_uv = props["psi_uv"].to_value(u.radian)
+        else:
+            raise RuntimeError(f"focalplane[{det}] does not include psi_uv")
+        return psipol
+
+    def _get_epsilon(self, focalplane, det):
+        """Parse polarization leakage (epsilon) from the focalplane
+        object or dictionary.
+        """
+        if det not in focalplane:
+            raise RuntimeError(f"focalplane does not include {det}")
+        props = focalplane[det]
+        if "pol_leakage" in props.colnames:
+            epsilon = focalplane[det]["pol_leakage"]
+        else:
+            # Assume zero polarization leakage
+            epsilon = 0
+        return epsilon
+
+    def load_alm(self, file, mmax):
+        def read_comp(file, comp, out):
+            almX, mmaxX = hp.fitsfunc.read_alm(file, hdu=comp+1, return_mmax=True)
+            lmaxX = hp.sphtfunc.Alm.getlmax(almX.shape[0], mmaxX)
+
+            ofs1, ofs2 = 0, 0
+            mylmax = min(self.lmax, lmaxX)
+            for m in range(0, min(mmax, mmaxX)+1):
+                out[comp,ofs1:ofs1+mylmax-m+1] = almX[ofs2:ofs2+mylmax-m+1]
+                ofs1 += self.lmax-m+1
+                ofs2 += lmaxX-m+1
+
+        ncomp = 3 if self.pol else 1
+        res = np.zeros((ncomp, hp.sphtfunc.Alm.getsize(self.lmax, mmax)), dtype=np.complex128)
+        for i in range(ncomp):
+            read_comp(file, i, res)
+        return res
+
+    def get_sky(self, skyfile, det, verbose):
+        timer = Timer()
+        timer.start()
+        sky = self.load_alm(skyfile, self.lmax)
+        fwhm = self.fwhm.to_value(u.radian)
+        # Hmm, healpy doesn't support un-smoothing via negative FWHM, so let's do it by hand ...
+        if fwhm != 0:
+            sigma = fwhm / (2.0 * np.sqrt(2.0 * np.log(2.0)))
+            for i in range(sky.shape[0]):            
+                ell = np.arange(self.lmax + 1.0)
+                s = 2 if i >= 1 else 0
+                fact = np.exp(0.5 * (ell * (ell + 1) - s ** 2) * sigma ** 2)
```

Comment:
You could just use `healpy.gauss_beam()` to get the window function. I presume you would want to use the proper polarized definition.

Reply:
Good point, thanks! In the longer term we should probably get this sort of deconvolution directly into `healpy`.
I think I should get the polarization right via the use of `s` (actually this is copy/paste from from the `healpy` sources with the sign of the exponential flipped).

Reply:
Fixed. This version is really much nicer!